### PR TITLE
gh-117225: Add color to doctest output

### PR DESCRIPTION
**Pull request by hugovk**
The doctest output looks like:

image

I moved the existing `_ANSIColors` class and `_can_colorize` function from `traceback.py`  added in https://github.com/python/cpython/pull/112732, to re-use them, cc @pablogsal.

* Is `_colorize` an okay name for this module?

* If the module has leading underscore, do we also want leading underscores in `_ANSIColors` and `_can_colorize`?

* Issue: gh-117225

### CODE REVIEW

File: Lib/_colorize.py
Comment (AlexWaygood) [inline Lib/_colorize.py]: Nit: since this is now a private module, I think I'd give all these public names. `from _colorize import _can_colorize` feels pretty ugly to me, and I think it's still obvious that it's private if it's called `_colorize.can_colorize`

```suggestion
COLORIZE = True


class ANSIColors:
    BOLD_GREEN = "\x1b[1;32m"
    BOLD_MAGENTA = "\x1b[1;35m"
    BOLD_RED = "\x1b[1;31m"
    GREEN = "\x1b[32m"
    GREY = "\x1b[90m"
    MAGENTA = "\x1b[35m"
    RED = "\x1b[31m"
    RESET = "\x1b[0m"
    YELLOW = "\x1b[33m"


def can_colorize():
```

File: Lib/test/test_doctest/test_doctest.py
Comment (AlexWaygood) [inline Lib/test/test_doctest/test_doctest.py]: Nit: do we need to add all the blank lines in this file? It would be nice to minimise the diff a little, I think

File: Lib/test/test_doctest/test_doctest.py
Comment (AlexWaygood) [inline Lib/test/test_doctest/test_doctest.py]: British spelling of "colour" :D

File: Lib/doctest.py
Comment (AlexWaygood) [inline Lib/doctest.py]: Perhaps a somewhat redundant comment ;) it's the `doctest` module, afterall

```suggestion
import _colorize
```

Also, any reason this import is in a separate block from the other imports above it? Everything's the stdlib here ;)

File: Lib/doctest.py
Comment (AlexWaygood) [inline Lib/doctest.py]: I think I'd rather we left this file in violation of PEP 8, so that we have a clean diff and it's really clear from the git history exactly which lines we needed to change to get colourised doctests

**Review by AlexWaygood — COMMENTED** [review]
Nice! A few nits below -- it's a little hard to read the overall diff because of all the blank lines being added. Would you mind maybe leaving those out for now?

**Comment by aisk** [comment]
I think we should also change the British spelling 'colour' to 'color' in the PR's title, which will become the commit message after merging.

File: Lib/doctest.py
Comment (hugovk) [reply Lib/doctest.py]: Sure, will revert that commit.

File: Lib/doctest.py
Comment (hugovk) [reply Lib/doctest.py]: I followed the pattern of this, I think the comment is because it can show up as unused to tools:

https://github.com/python/cpython/blob/62aeb0ee69b06091396398de56dcb755ca3b9dc9/Lib/test/test_doctest/test_doctest.py#L7

But can remove if you prefer? And remove the existing one for `FakeInput`?

Comment (AlexWaygood) [reply Lib/doctest.py]: Ahhh I misunderstood. The comment can stay, thanks for explaining!

File: Lib/traceback.py
Comment (AlexWaygood) [inline Lib/traceback.py]: Why not import this at the module level along with `ANSIColors`?

Comment (hugovk) [reply Lib/traceback.py]: Will move it.

Was originally thinking it's only used in this one place, so don't need to import most of the time. But then couldn't do the same with `ANSIColors`, so they might as well both be together at the top.

File: Lib/doctest.py
Comment (AlexWaygood) [inline Lib/doctest.py]: I would also move these imports to the top of the module, since we already have `import _colorize` at the top (for the doctests).

File: Lib/doctest.py
Comment (AlexWaygood) [inline Lib/doctest.py]: nit (sorry): I'd leave this variable name as it is for now, so that it's really clear from the diff for future readers exactly what needed to change to add colour support here

File: Lib/doctest.py
Comment (AlexWaygood) [inline Lib/doctest.py]: same here -- I think it's fine to switch to f-strings if we're touching the line anyway, but doing it to unrelated lines as part of this PR just makes the diff confusing, I feel :(

**Comment by AlexWaygood** [comment]
What about also colorising the divider line that's used to separate failed examples? So it would look like this?

image

This patch does it:

```diff
diff --git a/Lib/doctest.py b/Lib/doctest.py
index 00ff1a107c..80d0b70592 100644
--- a/Lib/doctest.py
+++ b/Lib/doctest.py
@@ -1307,7 +1307,10 @@ def report_unexpected_exception(self, out, test, example, exc_info):
             'Exception raised:\n' + _indent(_exception_traceback(exc_info)))
 
     def _failure_header(self, test, example):
-        out = [self.DIVIDER]
+        red, reset = (
+            (ANSIColors.RED, ANSIColors.RESET) if can_colorize() else ("", "")
+        )
+        out = [f"{red}{self.DIVIDER}{reset}"]
         if test.filename:
             if test.lineno is not None and example.lineno is not None:
                 lineno = test.lineno + example.lineno + 1
@@ -1621,7 +1624,7 @@ def summarize(self, verbose=None):
                     print(f" {green}{count:3d} test{s} in {name}{reset}")
 
         if failed:
-            print(self.DIVIDER)
+            print(f"{red}{self.DIVIDER}{reset}")
             print(f"{red}{_n_items(failed)} had failures:{reset}")
```

**Comment by AlexWaygood** [comment]
Of these three summary lines at the end:

```
1 item had failures:
   2 of   4 in typing._should_unflatten_callable_args
***Test Failed*** 2 failures.
```

I think I might be inclined to keep the last line colored in red (as you have it), but _not_ colorize the firsrt two of those lines. I feel like having everything in red makes it harder to read, and having the last line in red is enough to draw your attention to it and communicate "bad!". What do you think?

**Comment by hugovk** [comment]
I've added red divider lines.

> Of these three summary lines at the end:
> 
> ```
> 1 item had failures:
>    2 of   4 in typing._should_unflatten_callable_args
> ***Test Failed*** 2 failures.
> ```
> 
> You currently have all three colored in red if there are errors. I think I might be inclined to keep the last line colored in red (as you have it), but _not_ colorize the first two of those lines. I feel like having everything in red makes it harder to read, and having the last line in red is enough to draw your attention to it and communicate "bad!". What do you think?

Can you share a screenshot and a diff to trigger it?

I don't get all in red:

<details>
<summary>Screenshots</summary>

image

image


</details>

**Comment by AlexWaygood** [comment]
> Can you share a screenshot and a diff to trigger it?

Screenshot in https://github.com/python/cpython/pull/117583#issuecomment-2041223794. For how to trigger it — I was testing by doing some silly things to doctests Lib/typing.py, then running `./python.exe -m doctest Lib/typing.py` directly!

**Comment by hugovk** [comment]
I've pushed that, how does this look?

<details>
<summary>Screenshots</summary>

<table>
<tr>
 <th>Before
 <th>After
<tr>
 <td>
image
 <td>
image
<tr>
 <td>
image

 <td>
image

<tr>
 <td>
image

 <td>
image

</table>

</details>

**Comment by AlexWaygood** [comment]
> I've pushed that, how does this look?

Looks fantastic :D

File: Lib/test/test_doctest/test_doctest2.py
Comment (AlexWaygood) [inline Lib/test/test_doctest/test_doctest2.py]: It looks like tests pass for me locally if I delete this -- what's it for, exactly?

Comment (hugovk) [reply Lib/test/test_doctest/test_doctest2.py]: Hmm, same here... Reverted!

**Comment by Pblxyz** [comment]
#

File: Lib/test/test_traceback.py
Comment (AlexWaygood) [inline Lib/test/test_traceback.py]: Should these tests be moved to a new `test__colorize.py` file, now that the function they're testing no longer lives in `traceback.py`?

Comment (hugovk) [reply Lib/test/test_traceback.py]: Sure, done, with some light refactoring.

File: Lib/test/test__colorize.py
Comment (AlexWaygood) [inline Lib/test/test__colorize.py]: Sorry for not being clear :(

I think basically _just_ this test should go in `test__colorize.py`, as this is the only one that directly tests `_colorize.py`. ISTM that the other tests are asserting the _use_ of the `_colorize` module by the `traceback` module -- so probably those tests should stay in `test_traceback.py`

**Comment by AlexWaygood** [comment]
It looks like many tests fail if you run tests with `FORCE_COLOR`, e.g. `FORCE_COLOR ./python.exe -m test` (see https://github.com/python/cpython/issues/117605), and `test_doctest` is one of them. This patch does seem to make a few more parts of `test_doctest` fail with `FORCE_COLOR`, though:

image

Since there are so many that already fail with that environment variable set, I'm not sure it's really an issue, but thought I'd mention it anyway.

File: Lib/test/test__colorize.py
Comment (hugovk) [reply Lib/test/test__colorize.py]: No problem, updated!

File: Lib/test/test__colorize.py
Comment (AlexWaygood) [inline Lib/test/test__colorize.py]:
```suggestion
class TestColorizeFunction(unittest.TestCase):
```

File: Lib/test/test__colorize.py
Comment (AlexWaygood) [inline Lib/test/test__colorize.py]: If we use `subTest`, then it's nice and clear exactly which iteration of the `for` loop failed (if one does fail) -- the parameters passed to `subTest` are given in the `FAIL` header at the top:

```pytb
======================================================================
FAIL: test_colorized_detection_checks_for_environment_variables (test.test__colorize.TestColorizedTraceback.test_colorized_detection_checks_for_environment_variables) (env_vars={'PYTHON_COLORS': '1'}, expected_color=False)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/alexw/dev/cpython/Lib/test/test__colorize.py", line 37, in test_colorized_detection_checks_for_environment_variables
    self.assertEqual(_colorize.can_colorize(), expected)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: True != False

----------------------------------------------------------------------
```

```suggestion
                    with self.subTest(env_vars=env_vars, expected_color=expected):
                        with unittest.mock.patch("os.environ", env_vars):
                            self.assertEqual(_colorize.can_colorize(), expected)
```

File: Lib/test/test__colorize.py
Comment (AlexWaygood) [inline Lib/test/test__colorize.py]: Oh, actually, I think you can simplify this test quite a bit if you use the newish [`enterContext`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.enterContext) unittest feature. You could write it like this:

```py
    def test_colorized_detection_checks_for_environment_variables(self):
        env_vars_expected = [
            ({'TERM': 'dumb'}, False),
            ({'PYTHON_COLORS': '1'}, True),
            ({'PYTHON_COLORS': '0'}, False),
            ({'NO_COLOR': '1'}, False),
            ({'NO_COLOR': '1', "PYTHON_COLORS": '1'}, True),
            ({'FORCE_COLOR': '1'}, True),
            ({'FORCE_COLOR': '1', 'NO_COLOR': '1'}, False),
            ({'FORCE_COLOR': '1', "PYTHON_COLORS": '0'}, False),
        ]

        if sys.platform == "win32":
            self.enterContext(
                unittest.mock.patch(
                    "nt._supports_virtual_terminal", return_value=True
                )
            )

        isatty_mock = self.enterContext(unittest.mock.patch("os.isatty"))
        isatty_mock.return_value = True

        for env_vars, expected in env_vars_expected:
            with self.subTest(env_vars=env_vars, expected_color=expected):
                with unittest.mock.patch("os.environ", env_vars):
                    self.assertEqual(_colorize.can_colorize(), expected)

        isatty_mock.return_value = False
        self.assertEqual(_colorize.can_colorize(), False)
```

File: Lib/test/test__colorize.py
Comment (hugovk) [reply Lib/test/test__colorize.py]: Nice, I've not used `subTest` before, looks a bit like `pytest.mark.parametrize` 👍

**Review by AlexWaygood — APPROVED** [review]
Looks great to me! You might want to wait for a second review though (preferably from somebody who uses a light theme for their terminal 😄)

File: Lib/test/test__colorize.py
Comment (AlexWaygood) [reply Lib/test/test__colorize.py]: Subtests are great! In some ways, I actually prefer them to `parametrize` (though I much prefer pytest as a testing framework in general). There's a nice article on the differences between subtests and `parametrize` here: https://blog.ganssle.io/articles/2020/04/subtests-in-python.html

**Comment by Privat33r-dev** [comment]
I was pleased to see that `colorize` now follows the ISP. I love the idea of colorization, and I think that we can continue this trend with `unittests` as well. Thank you for your contribution!

The minor improvement that could be done is to use stub/other way to return empty colors so instead of this:
```py
        if can_colorize():
            bold_green = ANSIColors.BOLD_GREEN
            bold_red = ANSIColors.BOLD_RED
            green = ANSIColors.GREEN
            red = ANSIColors.RED
            reset = ANSIColors.RESET
            yellow = ANSIColors.YELLOW
        else:
            bold_green = ""
            bold_red = ""
            green = ""
            red = ""
            reset = ""
            yellow = ""
```

it could be something like this:
```py
        ANSIColors = get_colors()
        bold_green = ANSIColors.BOLD_GREEN
        bold_red = ANSIColors.BOLD_RED
        green = ANSIColors.GREEN
        red = ANSIColors.RED
        reset = ANSIColors.RESET
        yellow = ANSIColors.YELLOW
```

File: Lib/_colorize.py
Comment (Privat33r-dev) [inline Lib/_colorize.py]:
```suggestion
ANSIColorsStub = type('ANSIColorsStub', (ANSIColors,), {x: "" for x in dir(ANSIColors) if not x.startswith('__')})

def get_ansi_colors() -> ANSIColors:
  return ANSIColors if can_colorize() else ANSIColorsStub
```

Using `get_ansi_colors()` preserves IDE suggestions (ctrl+space) and centralizes, as well as abstracts color availability logic, improving code's clarity and flexibility.

File: Lib/_colorize.py
Comment (Privat33r-dev) [inline Lib/_colorize.py]:
```suggestion
def can_colorize() -> bool:
```

Consider adding typing.

**Review by Privat33r-dev — CHANGES_REQUESTED** [review]

File: Lib/_colorize.py
Comment (AlexWaygood) [reply Lib/_colorize.py]: In general our policy is currently not to add type hints to standard-library modules in the `Lib/` directory (there are some exceptions, but they're all pretty specific). I think we still have yet to add some formal documentation around this in the devguide (we really need to!), but you can see some rationales here:
- https://github.com/python/cpython/pull/108125#discussion_r1313654070
- https://discuss.python.org/t/type-annotations-in-the-stdlib/21487

Comment (Privat33r-dev) [reply Lib/_colorize.py]: Thank you for this explanation, I made another issue to formalize the problem. 

-------
I am not sure what technically prevents me to mark this part as `resolved`, but it can be considered resolved and marked as such by the author.

Comment (AlexWaygood) [reply Lib/_colorize.py]: Thanks so much @Privat33r-dev!

**Comment by hugovk** [comment]
Thanks for the suggestion, something like that sounds good.

I'll wait to give https://github.com/python/cpython/pull/117672 a change to be merged first so as not to cause conflicts there, then will update this. I might include the refactoring as part of this PR, but mainly want to make sure this is merged before the [beta cutoff on 2024-05-07](https://peps.python.org/pep-0719/#schedule) so it can be in 3.13.

I'll also add some docs, but they can be another PR and can be after the beta cutoff, if necessary.

File: Lib/traceback.py
Comment (AlexWaygood) [inline Lib/traceback.py:452]: Oh, is there a reason why these colours might have been omitted from the original changes to colourise Python's tracebacks in general? Maybe to avoid accessibility issues for people who are red-green colourblind? Or to avoid issues for people who have light themes on their terminals?

**Review by AlexWaygood — COMMENTED** [review]
The patch still looks good to me, but I'd still love it if we could get a review from somebody who uses a light theme in their terminal and/or somebody who's colourblind before we merge 👍

File: Lib/traceback.py
Comment (hugovk) [reply Lib/traceback.py:452]: I think because when there's a traceback error there's nothing "good" to report in green :)

**Comment by hugovk** [comment]
The widely-used pytest (116m monthly downloads) uses the same ANSI colours:

https://github.com/pytest-dev/pytest/blob/58844247f7ae4a7a213a73c8af2462253b3d8fc7/src/_pytest/_io/terminalwriter.py#L44-L66

@encukou made a good point in https://github.com/python/cherry-picker/pull/120#issuecomment-1960942870:

> The 16 named colours are determined by the terminal. If there's e.g. not enough contrast between background and red in the default settings, you should probably report it to the terminal devs.
> 
> That also means you shouldn't use direct RGB values, or the 256-colour palette. (Unless you also paint the background using RGB, which looks bad if it's not a “full-screen” app.)
> 
> So, my suggestion: stick to 16 named colours, implement `NO_COLOR` & `--color/--no-color` to opt out, but don't over-analyze how _your_ terminal renders things.

We're sticking to the 16 ANSI colours and have `PYTHON_COLORS` and `NO_COLOR` to disable.

@pablogsal How does the following look for you? Is the text still legible?

image

**Comment by hugovk** [comment]
Docs PR: https://github.com/python/cpython/pull/118268

Refactoring PR: https://github.com/python/cpython/pull/118283

**Pull request merged**
